File: 프로그래머스/2/42888. 오픈채팅방/오픈채팅방.java
import java.util.*;

class Solution {
    public String[] solution(String[] record) {

        //leave 하다보니 메세지도 저장할 필요성을 느낌
        List<String[]> list = new ArrayList<>();
        HashMap<String, String> user = new HashMap<>();
        //사용자 아이디 저장
        for (int i = 0; i < record.length; i++) {
            String[] arr = record[i].split(" ");
            //들어옴
            if(arr[0].equals("Enter")) {
                user.put(arr[1], arr[2]);
                list.add(new String[] { "Enter", arr[1] });
            }
            //ID 바꿈
            else if(arr[0].equals("Change")) {
                user.replace(arr[1], arr[2]);
            }
            //떠남
            else {
                list.add(new String[] { "Leave", arr[1] });
            }
        }

        //메세지 출력
        String[] answer = new String[list.size()];
        for (int i = 0; i < list.size(); i++) {
            String[] result = list.get(i);
            //result[0] : 명령어 뭐였는지
            //result[1] : userId
            //user.get(id) : id로 이름 가져옴
            if(result[0].equals("Enter")) {
                answer[i] = user.get(result[1]) + "님이 들어왔습니다.";
            }
            else if(result[0].equals("Leave")) {
                answer[i] = user.get(result[1]) + "님이 나갔습니다.";
            }
            else {
                continue;
            }
        }

        return answer;
    }
}

File: 프로그래머스/2/42888. 오픈채팅방/오픈채팅방.cpp
#include <iostream>
#include <vector>
#include <map>
#include <string>
#include <sstream>
 
using namespace std;

vector<string> solution(vector<string> record) {
    vector<string> answer;
    map<string, string> user;
    //map에 명령어 저장
    for (const auto& line : record) {
        stringstream ss(line);
        string message, id, nickname;
        ss >> message >> id;
       
        if (message == "Enter") {
            ss >> nickname;
            user[id] = nickname;
        }
        else if (message == "Change") {
            ss >> nickname;
            user[id] = nickname;
        }
    }

    for (const auto& line : record) {
        stringstream ss(line);
        string cmd, id;
        ss >> cmd >> id;
        if (cmd == "Enter") {
            answer.push_back(user[id] + "님이 들어왔습니다.");
        }
        else if (cmd == "Leave") {
            answer.push_back(user[id] + "님이 나갔습니다.");
        }
    }
    return answer;
}